File: python/130.surrounded-regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        ROWS, COLS = len(board), len(board[0])
        DIRECTIONS = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        def dfs_explore(r, c):
            if (r < 0 or r >= ROWS or 
                c < 0 or c >= COLS or 
                board[r][c] != 'O'):
                return 
            
            board[r][c] = 'D'
            for dr, dc in DIRECTIONS:
                dfs_explore(r + dr, c + dc)

        for r in range(ROWS):
            if board[r][0] == 'O':  
                dfs_explore(r, 0)
            if board[r][COLS - 1] == 'O':  
                dfs_explore(r, COLS - 1)

        for c in range(COLS):
            if board[0][c] == 'O':  
                dfs_explore(0, c)
            if board[ROWS - 1][c] == 'O':  
                dfs_explore(ROWS - 1, c)

        for row in range(ROWS):
            for col in range(COLS):
                if board[row][col] == 'O':
                    board[row][col] = 'X' 
                    
                elif board[row][col] == 'D':
                    board[row][col] = 'O' 

File: go/130.surrounded-regions.go

func solve(board [][]byte)  {

    m := len(board)
    n := len(board[0])
 
 /*
    print := func(){
        for _, subarr := range (board) {
            for _, elt := range (subarr) {
                fmt.Printf("%c ", elt)
            }
        fmt.Println()
        }
        fmt.Println()
    }
    */

   var dfs func(int, int)
    dfs = func(i, j int) {
        if i < 0 || j < 0 || i >= m || j >= n || board[i][j] == 'X' || board[i][j] == '*' {
            return
        }

        board[i][j] = '*'
        dfs(i-1, j)
        dfs(i+1, j)
        dfs(i, j-1)
        dfs(i, j+1)
    }

    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            if i == 0 || j == 0 || i == m - 1 || j == n - 1 {
                if board[i][j] == 'O' {
                    dfs(i, j)
                }
            }
        }
    }
    
    for i := 0; i < m; i++ {
        for j := 0; j < n; j++ {
            if board[i][j] == '*' {
                board[i][j] = 'O'
            } else if board[i][j] == 'O' {
                board[i][j] = 'X'
            }
        }
    }    
}